File: functions/services/keyword-scorer.js
/**
 * services/keyword-scorer.js
 * 키워드 점수 계산 및 품질 분석
 */

'use strict';

/**
 * SERP 점수 계산 (검색 결과 페이지 분석)
 * @param {Object} serpData - SERP 분석 데이터
 * @returns {number} 0-10 점수
 */
function calculateSERPScore(serpData) {
  if (!serpData || !serpData.results || serpData.results.length === 0) {
    return 5; // 기본 중립 점수
  }

  let score = 10; // 최고 점수에서 시작

  const { blogCount, officialCount, totalResults } = serpData;

  // 1. 공식 사이트 패널티
  // .go.kr, .or.kr, 뉴스 사이트가 많으면 경쟁이 어려움
  const officialRatio = totalResults > 0 ? officialCount / totalResults : 0;

  if (officialRatio > 0.6) {
    score -= 4; // 60% 이상 공식 사이트면 큰 패널티
  } else if (officialRatio > 0.4) {
    score -= 3;
  } else if (officialRatio > 0.2) {
    score -= 2;
  }

  // 2. 블로그 비율 보너스
  // 블로그가 많으면 개인도 경쟁 가능
  const blogRatio = totalResults > 0 ? blogCount / totalResults : 0;

  if (blogRatio > 0.6) {
    score += 3; // 60% 이상 블로그면 큰 보너스
  } else if (blogRatio > 0.4) {
    score += 2;
  } else if (blogRatio > 0.2) {
    score += 1;
  }

  // 3. 오래된 콘텐츠 보너스
  // (날짜 정보가 있다면 분석)
  // 여기서는 간단히 처리, 실제로는 serpData.results에서 날짜 추출 필요

  // 점수 범위 제한
  return Math.max(0, Math.min(10, score));
}

/**
 * 경쟁도 점수 계산
 * @param {number} resultCount - 검색 결과 수
 * @returns {number} 0-10 점수
 */
function calculateCompetitionScore(resultCount) {
  // 검색 결과가 적을수록 경쟁이 낮음 = 높은 점수
  if (resultCount < 100) {
    return 10;
  } else if (resultCount < 500) {
    return 9;
  } else if (resultCount < 1000) {
    return 8;
  } else if (resultCount < 5000) {
    return 7;
  } else if (resultCount < 10000) {
    return 6;
  } else if (resultCount < 50000) {
    return 5;
  } else if (resultCount < 100000) {
    return 4;
  } else if (resultCount < 500000) {
    return 3;
  } else if (resultCount < 1000000) {
    return 2;
  } else {
    return 1;
  }
}

/**
 * 구체성 점수 계산 (단어 수 기반)
 * @param {string} keyword - 키워드
 * @returns {number} 0-10 점수
 */
function calculateSpecificityScore(keyword) {
  // 띄어쓰기로 단어 수 계산
  const words = keyword.trim().split(/\s+/);
  const wordCount = words.length;

  // 롱테일 키워드일수록 높은 점수
  if (wordCount >= 5) {
    return 10;
  } else if (wordCount === 4) {
    return 9;
  } else if (wordCount === 3) {
    return 7;
  } else if (wordCount === 2) {
    return 5;
  } else {
    return 3; // 단일 단어는 경쟁이 너무 높음
  }
}

/**
 * 블로그 비율 점수 계산
 * @param {Object} serpData - SERP 분석 데이터
 * @returns {number} 0-10 점수
 */
function calculateBlogRatioScore(serpData) {
  if (!serpData || !serpData.totalResults || serpData.totalResults === 0) {
    return 5;
  }

  const blogRatio = serpData.blogCount / serpData.totalResults;

  // 블로그 비율이 높을수록 높은 점수
  if (blogRatio >= 0.8) {
    return 10;
  } else if (blogRatio >= 0.6) {
    return 9;
  } else if (blogRatio >= 0.4) {
    return 7;
  } else if (blogRatio >= 0.2) {
    return 5;
  } else {
    return 3;
  }
}

/**
 * 정치 관련성 점수 계산
 * @param {string} keyword - 키워드
 * @param {string} district - 지역구
 * @param {string} topic - 주제
 * @returns {number} 0-10 점수
 */
function calculateRelevanceScore(keyword, district, topic) {
  let score = 5; // 기본 점수

  const lowerKeyword = keyword.toLowerCase();

  // 지역구 포함 여부
  if (district && lowerKeyword.includes(district.toLowerCase())) {
    score += 3;
  }

  // 주제 포함 여부
  if (topic && lowerKeyword.includes(topic.toLowerCase())) {
    score += 2;
  }

  // 정치 관련 키워드
  const politicalKeywords = [
    '의원', '국회', '시의회', '구의회', '정책', '공약',
    '지역', '주민', '시민', '민생', '복지', '개발',
    '예산', '조례', '의정', '활동', '행정', '사업'
  ];

  const hasPoliticalKeyword = politicalKeywords.some(pk =>
    lowerKeyword.includes(pk)
  );

  if (hasPoliticalKeyword) {
    score += 2;
  }

  return Math.min(10, score);
}

/**
 * 최종 점수 계산 (가중 평균)
 * @param {Object} scores - 각 항목별 점수
 * @returns {number} 최종 점수 (0-100)
 */
function calculateFinalScore(scores) {
  const {
    competitionScore = 5,
    specificityScore = 5,
    blogRatioScore = 5,
    trendScore = 5,
    relevanceScore = 5
  } = scores;

  // 가중치
  const weights = {
    competition: 0.35,    // 35%
    specificity: 0.25,    // 25%
    blogRatio: 0.20,      // 20%
    trend: 0.10,          // 10%
    relevance: 0.10       // 10%
  };

  // 가중 평균 계산
  const finalScore =
    (competitionScore * weights.competition) +
    (specificityScore * weights.specificity) +
    (blogRatioScore * weights.blogRatio) +
    (trendScore * weights.trend) +
    (relevanceScore * weights.relevance);

  // 0-100 범위로 변환
  return Math.round(finalScore * 10);
}

/**
 * 키워드 등급 부여
 * @param {number} finalScore - 최종 점수
 * @returns {string} 등급 (S, A, B, C, D)
 */
function getKeywordGrade(finalScore) {
  if (finalScore >= 85) return 'S'; // 최상급
  if (finalScore >= 70) return 'A'; // 상급
  if (finalScore >= 55) return 'B'; // 중상급
  if (finalScore >= 40) return 'C'; // 중급
  return 'D'; // 하급
}

/**
 * 키워드 종합 분석
 * @param {Object} params - 분석 파라미터
 * @returns {Object} 종합 분석 결과
 */
function analyzeKeyword(params) {
  const {
    keyword,
    serpData,
    resultCount,
    trendScore,
    district,
    topic
  } = params;

  // 각 항목별 점수 계산
  const scores = {
    competitionScore: calculateCompetitionScore(resultCount),
    specificityScore: calculateSpecificityScore(keyword),
    blogRatioScore: calculateBlogRatioScore(serpData),
    trendScore: trendScore || 5,
    relevanceScore: calculateRelevanceScore(keyword, district, topic),
    serpScore: calculateSERPScore(serpData)
  };

  // 최종 점수
  const finalScore = calculateFinalScore(scores);
  const grade = getKeywordGrade(finalScore);

  // 추천 이유 생성
  const reasons = generateRecommendationReasons(scores, finalScore);

  return {
    keyword,
    finalScore,
    grade,
    scores,
    reasons,
    metadata: {
      resultCount,
      blogRatio: serpData ? (serpData.blogCount / serpData.totalResults) : 0,
      officialRatio: serpData ? (serpData.officialCount / serpData.totalResults) : 0
    }
  };
}

/**
 * 추천 이유 생성
 * @param {Object} scores - 점수 객체
 * @param {number} finalScore - 최종 점수
 * @returns {Array<string>} 추천 이유 배열
 */
function generateRecommendationReasons(scores, finalScore) {
  const reasons = [];

  if (scores.competitionScore >= 8) {
    reasons.push('🎯 경쟁이 낮아 상위 노출 가능성이 높습니다');
  }

  if (scores.specificityScore >= 7) {
    reasons.push('📝 구체적인 롱테일 키워드로 타겟팅이 명확합니다');
  }

  if (scores.blogRatioScore >= 7) {
    reasons.push('📰 블로그 콘텐츠가 많아 개인도 경쟁 가능합니다');
  }

  if (scores.trendScore >= 8) {
    reasons.push('📈 검색량이 증가하는 트렌드입니다');
  }

  if (scores.relevanceScore >= 8) {
    reasons.push('🏛️ 정치/지역 관련성이 매우 높습니다');
  }

  if (finalScore >= 85) {
    reasons.push('⭐ 최상위 등급 키워드입니다');
  }

  if (reasons.length === 0) {
    reasons.push('💡 일반적인 키워드입니다');
  }

  return reasons;
}

/**
 * 키워드 배치 분석
 * @param {Array<Object>} keywords - 키워드 배열 (각각 분석 파라미터 포함)
 * @returns {Array<Object>} 분석 결과 배열 (점수순 정렬)
 */
function analyzeKeywordBatch(keywords) {
  console.log(`📊 [Scorer] 배치 분석 시작: ${keywords.length}개 키워드`);

  const results = keywords.map(params => analyzeKeyword(params));

  // 최종 점수 기준 내림차순 정렬
  results.sort((a, b) => b.finalScore - a.finalScore);

  console.log(`✅ [Scorer] 배치 분석 완료`);

  return results;
}

module.exports = {
  calculateSERPScore,
  calculateCompetitionScore,
  calculateSpecificityScore,
  calculateBlogRatioScore,
  calculateRelevanceScore,
  calculateFinalScore,
  getKeywordGrade,
  analyzeKeyword,
  analyzeKeywordBatch
};

File: functions/python/services/keyword_analysis/keyword_scorer.py
"""
Keyword scoring service migrated from `functions/services/keyword-scorer.js`.
"""

from __future__ import annotations

from typing import Any, Dict, List


def _safe_list(value: Any) -> List[Any]:
    return value if isinstance(value, list) else []


def _safe_number(value: Any, default: float = 0.0) -> float:
    try:
        if value is None:
            return default
        if isinstance(value, bool):
            return float(int(value))
        if isinstance(value, (int, float)):
            return float(value)
        text = str(value).strip()
        if not text:
            return default
        return float(text)
    except Exception:
        return default


def calculate_serp_score(serp_data: Dict[str, Any]) -> float:
    if not isinstance(serp_data, dict):
        return 5

    results = _safe_list(serp_data.get("results"))
    if len(results) == 0:
        return 5

    score = 10.0
    blog_count = _safe_number(serp_data.get("blogCount"), 0)
    official_count = _safe_number(serp_data.get("officialCount"), 0)
    total_results = _safe_number(serp_data.get("totalResults"), 0)

    official_ratio = (official_count / total_results) if total_results > 0 else 0
    if official_ratio > 0.6:
        score -= 4
    elif official_ratio > 0.4:
        score -= 3
    elif official_ratio > 0.2:
        score -= 2

    blog_ratio = (blog_count / total_results) if total_results > 0 else 0
    if blog_ratio > 0.6:
        score += 3
    elif blog_ratio > 0.4:
        score += 2
    elif blog_ratio > 0.2:
        score += 1

    return max(0, min(10, score))


def calculate_competition_score(result_count: Any) -> int:
    count = int(_safe_number(result_count, 0))
    if count < 100:
        return 10
    if count < 500:
        return 9
    if count < 1000:
        return 8
    if count < 5000:
        return 7
    if count < 10000:
        return 6
    if count < 50000:
        return 5
    if count < 100000:
        return 4
    if count < 500000:
        return 3
    if count < 1000000:
        return 2
    return 1


def calculate_specificity_score(keyword: str) -> int:
    word_count = len([part for part in str(keyword or "").strip().split() if part.strip()])
    if word_count >= 5:
        return 10
    if word_count == 4:
        return 9
    if word_count == 3:
        return 7
    if word_count == 2:
        return 5
    return 3


def calculate_blog_ratio_score(serp_data: Dict[str, Any]) -> int:
    if not isinstance(serp_data, dict):
        return 5

    total_results = _safe_number(serp_data.get("totalResults"), 0)
    if total_results <= 0:
        return 5
    blog_count = _safe_number(serp_data.get("blogCount"), 0)
    blog_ratio = blog_count / total_results
    if blog_ratio >= 0.8:
        return 10
    if blog_ratio >= 0.6:
        return 9
    if blog_ratio >= 0.4:
        return 7
    if blog_ratio >= 0.2:
        return 5
    return 3


def calculate_relevance_score(keyword: str, district: str, topic: str) -> int:
    score = 5
    lower_keyword = str(keyword or "").lower()
    lower_district = str(district or "").lower()
    lower_topic = str(topic or "").lower()

    if lower_district and lower_district in lower_keyword:
        score += 3
    if lower_topic and lower_topic in lower_keyword:
        score += 2

    political_keywords = [
        "\uC758\uC6D0",  # 의원
        "\uAD6D\uD68C",  # 국회
        "\uC2DC\uC758\uD68C",  # 시의회
        "\uAD6C\uC758\uD68C",  # 구의회
        "\uC815\uCC45",  # 정책
        "\uACF5\uC57D",  # 공약
        "\uC9C0\uC5ED",  # 지역
        "\uC8FC\uBBFC",  # 주민
        "\uBBFC\uC0DD",  # 민생
        "\uBCF5\uC9C0",  # 복지
        "\uAC1C\uBC1C",  # 개발
        "\uC608\uC0B0",  # 예산
        "\uC870\uB840",  # 조례
        "\uD589\uC815",  # 행정
        "\uD65C\uB3D9",  # 활동
        "\uC0AC\uC5C5",  # 사업
    ]
    if any(token in lower_keyword for token in political_keywords):
        score += 2

    return min(10, score)


def calculate_final_score(scores: Dict[str, Any]) -> int:
    competition_score = _safe_number(scores.get("competitionScore"), 5)
    specificity_score = _safe_number(scores.get("specificityScore"), 5)
    blog_ratio_score = _safe_number(scores.get("blogRatioScore"), 5)
    trend_score = _safe_number(scores.get("trendScore"), 5)
    relevance_score = _safe_number(scores.get("relevanceScore"), 5)

    final_score = (
        (competition_score * 0.35)
        + (specificity_score * 0.25)
        + (blog_ratio_score * 0.20)
        + (trend_score * 0.10)
        + (relevance_score * 0.10)
    )
    return int(round(final_score * 10))


def get_keyword_grade(final_score: int) -> str:
    if final_score >= 85:
        return "S"
    if final_score >= 70:
        return "A"
    if final_score >= 55:
        return "B"
    if final_score >= 40:
        return "C"
    return "D"


def generate_recommendation_reasons(scores: Dict[str, Any], final_score: int) -> List[str]:
    reasons: List[str] = []

    if _safe_number(scores.get("competitionScore"), 0) >= 8:
        reasons.append("경쟁도가 낮아 상위 노출 가능성이 높습니다")
    if _safe_number(scores.get("specificityScore"), 0) >= 7:
        reasons.append("구체적인 롱테일 키워드입니다")
    if _safe_number(scores.get("blogRatioScore"), 0) >= 7:
        reasons.append("블로그 콘텐츠 비중이 높아 개인 블로그에 유리합니다")
    if _safe_number(scores.get("trendScore"), 0) >= 8:
        reasons.append("검색량이 상승 중인 트렌드 키워드입니다")
    if _safe_number(scores.get("relevanceScore"), 0) >= 8:
        reasons.append("지역/정치 연관성이 높습니다")
    if final_score >= 85:
        reasons.append("최상위 등급의 키워드입니다")
    if len(reasons) == 0:
        reasons.append("일반적인 키워드입니다")

    return reasons


def analyze_keyword(params: Dict[str, Any]) -> Dict[str, Any]:
    keyword = str(params.get("keyword") or "")
    serp_data = params.get("serpData") if isinstance(params.get("serpData"), dict) else {}
    result_count = params.get("resultCount")
    trend_score = _safe_number(params.get("trendScore"), 5)
    district = str(params.get("district") or "")
    topic = str(params.get("topic") or "")

    scores: Dict[str, Any] = {
        "competitionScore": calculate_competition_score(result_count),
        "specificityScore": calculate_specificity_score(keyword),
        "blogRatioScore": calculate_blog_ratio_score(serp_data),
        "trendScore": trend_score,
        "relevanceScore": calculate_relevance_score(keyword, district, topic),
        "serpScore": calculate_serp_score(serp_data),
    }

    final_score = calculate_final_score(scores)
    grade = get_keyword_grade(final_score)
    reasons = generate_recommendation_reasons(scores, final_score)
    total_results = _safe_number(serp_data.get("totalResults"), 0)

    return {
        "keyword": keyword,
        "finalScore": final_score,
        "grade": grade,
        "scores": scores,
        "reasons": reasons,
        "metadata": {
            "resultCount": int(_safe_number(result_count, 0)),
            "blogRatio": (_safe_number(serp_data.get("blogCount"), 0) / total_results) if total_results > 0 else 0,
            "officialRatio": (_safe_number(serp_data.get("officialCount"), 0) / total_results) if total_results > 0 else 0,
        },
    }


def analyze_keyword_batch(keywords: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    results = [analyze_keyword(item if isinstance(item, dict) else {}) for item in keywords]
    results.sort(key=lambda item: int(item.get("finalScore") or 0), reverse=True)
    return results
